File: radix_sort/radix_sort.py
# Following is the implementation of Radix Sort.


def radix_sort(arr, radix=10):
    """
    :param arr: Iterable of elements to sort.
    :param radix: Base of input numbers
    :return: Sorted list of input.
    Time complexity: O(d * (n + b))
    where, n is the size of input list.
           b is base of representation.
           d is number of digits in largest number in that base.
    Space complexity: O(n + k)
    where, k is the range of input.
    """
    max_length = False
    tmp, digit = -1, 1

    while not max_length:
        max_length = True
        # declare and initialize buckets
        buckets = [[] for _ in range(radix)]

        # split arr between lists
        for i in arr:
            tmp = i // digit
            buckets[tmp % radix].append(i)
            if max_length and tmp > 0:
                max_length = False

        # empty lists into arr array
        a = 0
        for b in range(radix):
            buck = buckets[b]
            for i in buck:
                arr[a] = i
                a += 1

        # move to next digit
        digit *= radix
    return arr


def main():
    """
    Driver function to test radix sort.
    """
    test = [170, 45, 75, 90, 802, 24, 2, 66]
    print('Sorted array:', radix_sort(test))


if __name__ == '__main__':
    main()

File: radix_sort/RadixSort.java
/**
 * This Java program implements the Radix sort algorithm
 * It's a non-comparative based sorting algorithm, hence its worse case time
 * complexity is O(kn), and space is O(k + n) where k is the bucket size
 */

public class RadixSort {

    private static void radixSort(int[] array) {

        int m = array[0];
        int ex = 1;
        int n = array.length;
        int[] b = new int[10]; // initial bucket

        // loop through the array
        // find the max element
        for(int i = 1; i < n; i++) {
            if(array[i] > m)
                m = array[i];
        }

        while(m / ex > 0) {

            int[] bucket = new int[10];

            for(int i = 0; i < n; i++)
                bucket[(array[i] / ex) % 10]++;
            for(int i = 1; i < 10; i++)
                bucket[i] += bucket[i - 1];
            for(int i = n - 1; i >= 0; i--)
                b[--bucket[(array[i] / ex) % 10]] = array[i];

            for(int i = 0; i < n; i++)
                array[i] = b[i];
            ex *= 10;
        }
    }

    public static void main(String args[]) {
        int test[] = new int[]{170, 45, 75, 90, 802, 24, 2, 6};
        radixSort(test);

        for(int i: test) {
            System.out.print(i + " ");
        }
    }
}
